internal/terraform/rootmodule/root_module.go

```diff
@@ -394,9 +394,44 @@ func (rm *rootModule) ExecuteTerraformValidate(ctx context.Context) (map[string]
 	}
 
 	// an entry for each file should exist, even if there are no diags
-	for filename := range rm.pFilesMap {
+	for filename := range rm.parsedFiles() {
 		diagsMap[filename] = make(hcl.Diagnostics, 0)
 	}
+	// since validation applies to linked modules, create an entry for all
+	// files of linked modules
+	for _, m := range rm.moduleManifest.Records {
```

Comment:
Is there a reason for reading the module manifest here? Could we not create the entries "lazily" automatically below? e.g. something like

```go
diags, ok := diagsMap[d.Range.Filename]
if !ok {
    diagsMap[d.Range.Filename] = make(hcl.Diagnostics, 0)
    diags = diagsMap[d.Range.Filename]
}
```

I understand there is a need for filtering out submodule diags, but couldn't this be done via something like `strings.HasPrefix(d.Range.Filename, rm.Path())`?

Reply:
ah I don't believe so no, the point of this is to create an entry in the map for every possible file, so that clearing diagnostics over the protocol works correctly. I also authored this a while though so I can't describe it in detail, but we can revisit/refactor in follow ups